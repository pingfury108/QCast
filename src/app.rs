use async_trait::async_trait;
use loco_rs::{
    app::{AppContext, Hooks, Initializer},
    bgworker::Queue,
    boot::{create_app, BootResult, StartMode},
    config::Config,
    controller::AppRoutes,
    db::{self, truncate_table},
    environment::Environment,
    task::Tasks,
    Result,
};
use migration::Migrator;
use std::path::Path;

#[allow(unused_imports)]
use crate::{controllers, initializers, models::_entities::users, tasks, workers};

pub struct App;
#[async_trait]
impl Hooks for App {
    fn app_name() -> &'static str {
        env!("CARGO_CRATE_NAME")
    }

    fn app_version() -> String {
        format!(
            "{} ({})",
            env!("CARGO_PKG_VERSION"),
            option_env!("BUILD_SHA")
                .or(option_env!("GITHUB_SHA"))
                .unwrap_or("dev")
        )
    }

    async fn boot(
        mode: StartMode,
        environment: &Environment,
        config: Config,
    ) -> Result<BootResult> {
        create_app::<Self, Migrator>(mode, environment, config).await
    }

    async fn initializers(_ctx: &AppContext) -> Result<Vec<Box<dyn Initializer>>> {
        Ok(vec![])
    }

    fn routes(_ctx: &AppContext) -> AppRoutes {
        AppRoutes::with_default_routes() // controller routes below
            .add_route(controllers::books::routes())
            .add_route(controllers::chapters::routes())
            .add_route(controllers::medias::routes())
            .add_route(controllers::public::routes())
            .add_route(controllers::auth::routes())
            .add_route(controllers::dashboard::routes())
            // 后台管理路由
            .add_route(controllers::admin::users::routes())
            .add_route(controllers::admin::groups::routes())
    }
    async fn connect_workers(_ctx: &AppContext, _queue: &Queue) -> Result<()> {
        // 暂无 worker
        Ok(())
    }

    #[allow(unused_variables)]
    fn register_tasks(tasks: &mut Tasks) {
        tasks.register(tasks::create_superadmin::CreateSuperadmin);
        // tasks-inject (do not remove)
    }
    async fn truncate(ctx: &AppContext) -> Result<()> {
        truncate_table(&ctx.db, users::Entity).await?;
        Ok(())
    }
    async fn seed(ctx: &AppContext, base: &Path) -> Result<()> {
        db::seed::<users::ActiveModel>(&ctx.db, &base.join("users.yaml").display().to_string())
            .await?;
        Ok(())
    }
}
