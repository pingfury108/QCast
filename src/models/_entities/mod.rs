//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.16

pub mod prelude;

pub mod books;
pub mod chapters;
pub mod medias;
pub mod site_settings;
pub mod user_group_members;
pub mod user_groups;
pub mod users;
