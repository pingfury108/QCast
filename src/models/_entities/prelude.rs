//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.16

pub use super::books::Entity as Books;
pub use super::chapters::Entity as Chapters;
pub use super::medias::Entity as Medias;
pub use super::site_settings::Entity as SiteSettings;
pub use super::user_group_members::Entity as UserGroupMembers;
pub use super::user_groups::Entity as UserGroups;
pub use super::users::Entity as Users;
