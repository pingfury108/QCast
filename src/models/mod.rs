pub mod _entities;
pub mod books;
pub mod chapters;
pub mod medias;
pub mod user_group_members;
pub mod user_groups;
pub mod users;
