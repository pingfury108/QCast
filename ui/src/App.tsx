import { QueryClientProvider } from '@tanstack/react-query';
import { ReactQueryDevtools } from '@tanstack/react-query-devtools';
import { RouterProvider } from 'react-router-dom';
import { Toaster } from 'sonner';
import { queryClient } from './lib/query-client';
import { router } from './routes';

function App() {
  return (
    <QueryClientProvider client={queryClient}>
      <RouterProvider router={router} />
      <Toaster position="top-right" richColors />
      <ReactQueryDevtools initialIsOpen={false} />
    </QueryClientProvider>
  );
}

export default App;
