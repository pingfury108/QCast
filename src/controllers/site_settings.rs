#![allow(clippy::missing_errors_doc)]
#![allow(clippy::unnecessary_struct_initialization)]
#![allow(clippy::unused_async)]
use axum::debug_handler;
use loco_rs::prelude::*;
use serde::{Deserialize, Serialize};

use crate::models::{site_settings, users};

const DEFAULT_SITE_URL: &str = "http://localhost:5150";

#[derive(Debug, Serialize, Deserialize)]
pub struct SiteSettingsResponse {
    pub id: i32,
    pub site_url: String,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateSiteSettingsParams {
    pub site_url: String,
}

/// 获取站点设置
#[debug_handler]
pub async fn get_settings(auth: auth::JWT, State(ctx): State<AppContext>) -> Result<Response> {
    // 验证管理员权限
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    if !user.is_admin() {
        return unauthorized("Admin permission required");
    }

    // 获取或创建站点设置
    let settings = site_settings::Model::get_or_create(&ctx.db, DEFAULT_SITE_URL).await?;

    let response = SiteSettingsResponse {
        id: settings.id,
        site_url: settings.site_url,
        created_at: settings.created_at.into(),
        updated_at: settings.updated_at.into(),
    };

    format::json(response)
}

/// 更新站点设置
#[debug_handler]
pub async fn update_settings(
    auth: auth::JWT,
    State(ctx): State<AppContext>,
    Json(params): Json<UpdateSiteSettingsParams>,
) -> Result<Response> {
    // 验证管理员权限
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    if !user.is_admin() {
        return unauthorized("Admin permission required");
    }

    // 更新站点URL
    let settings = site_settings::Model::update_url(&ctx.db, params.site_url).await?;

    let response = SiteSettingsResponse {
        id: settings.id,
        site_url: settings.site_url,
        created_at: settings.created_at.into(),
        updated_at: settings.updated_at.into(),
    };

    format::json(response)
}

pub fn routes() -> Routes {
    Routes::new()
        .prefix("/api/admin/site-settings")
        .add("", get(get_settings))
        .add("", put(update_settings))
}
