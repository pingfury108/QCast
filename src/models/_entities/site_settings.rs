//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.16

use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "site_settings")]
pub struct Model {
    pub created_at: DateTimeWithTimeZone,
    pub updated_at: DateTimeWithTimeZone,
    #[sea_orm(primary_key)]
    pub id: i32,
    pub site_url: String,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}
