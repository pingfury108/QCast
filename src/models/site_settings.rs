use loco_rs::prelude::*;
use sea_orm::ActiveValue;
use serde::{Deserialize, Serialize};

pub use super::_entities::site_settings::{self, ActiveModel, Entity, Model};

#[derive(Debug, Deserialize, Serialize)]
pub struct UpdateSiteUrlParams {
    pub site_url: String,
}

impl Model {
    /// 获取站点设置，如果不存在则创建默认设置
    ///
    /// # Errors
    ///
    /// When database query fails
    pub async fn get_or_create(db: &DatabaseConnection, default_url: &str) -> ModelResult<Self> {
        // 尝试获取第一条记录（应该只有一条）
        if let Some(settings) = site_settings::Entity::find_by_id(1).one(db).await? {
            return Ok(settings);
        }

        // 如果不存在，创建默认设置
        let now = chrono::Utc::now();
        let settings = site_settings::ActiveModel {
            id: ActiveValue::Set(1),
            site_url: ActiveValue::Set(default_url.to_string()),
            created_at: ActiveValue::Set(now.into()),
            updated_at: ActiveValue::Set(now.into()),
        }
        .insert(db)
        .await?;

        Ok(settings)
    }

    /// 更新站点URL
    ///
    /// # Errors
    ///
    /// When database query fails or URL validation fails
    pub async fn update_url(db: &DatabaseConnection, new_url: String) -> ModelResult<Self> {
        // 验证URL格式
        validate_site_url(&new_url).map_err(|e| ModelError::Any(e.into()))?;

        // 获取或创建设置
        let settings = Self::get_or_create(db, &new_url).await?;

        // 更新URL
        let mut active: ActiveModel = settings.into();
        active.site_url = ActiveValue::Set(normalize_url(&new_url));
        active.update(db).await.map_err(ModelError::from)
    }
}

#[async_trait::async_trait]
impl ActiveModelBehavior for super::_entities::site_settings::ActiveModel {
    async fn before_save<C>(self, _db: &C, _insert: bool) -> Result<Self, DbErr>
    where
        C: ConnectionTrait,
    {
        // 验证 site_url
        if let ActiveValue::Set(ref url) = self.site_url {
            validate_site_url(url).map_err(DbErr::Custom)?;
        }
        Ok(self)
    }
}

/// 验证站点URL格式
fn validate_site_url(url: &str) -> Result<(), String> {
    if url.trim().is_empty() {
        return Err("Site URL cannot be empty".to_string());
    }

    // 必须以 http:// 或 https:// 开头
    if !url.starts_with("http://") && !url.starts_with("https://") {
        return Err("Site URL must start with http:// or https://".to_string());
    }

    // 验证是否是有效的URL
    url::Url::parse(url).map_err(|_| "Invalid URL format".to_string())?;

    Ok(())
}

/// 标准化URL（去除尾部斜杠）
fn normalize_url(url: &str) -> String {
    url.trim_end_matches('/').to_string()
}
