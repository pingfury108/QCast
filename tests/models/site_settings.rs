use insta::{assert_debug_snapshot, with_settings};
use loco_rs::testing::prelude::*;
use qcast::app::App;
use qcast::models::site_settings;
use serial_test::serial;

macro_rules! configure_insta {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        let _guard = settings.bind_to_scope();
    };
}

#[tokio::test]
#[serial]
async fn can_get_or_create_settings() {
    configure_insta!();

    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    let default_url = "http://localhost:5150";
    let settings = site_settings::Model::get_or_create(&boot.app_context.db, default_url)
        .await
        .unwrap();

    assert_eq!(settings.id, 1);
    assert_eq!(settings.site_url, default_url);

    with_settings!({
        filters => cleanup_site_settings_model()
    }, {
        assert_debug_snapshot!(settings);
    });
}

#[tokio::test]
#[serial]
async fn can_update_site_url() {
    configure_insta!();

    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    // 创建初始设置
    let _initial_settings =
        site_settings::Model::get_or_create(&boot.app_context.db, "http://localhost:5150")
            .await
            .unwrap();

    // 更新URL
    let new_url = "https://example.com";
    let updated_settings =
        site_settings::Model::update_url(&boot.app_context.db, new_url.to_string())
            .await
            .unwrap();

    assert_eq!(updated_settings.site_url, new_url);
    // 注意：由于更新太快，updated_at 可能等于 created_at，所以使用 >= 比较
    assert!(updated_settings.updated_at >= updated_settings.created_at);

    with_settings!({
        filters => cleanup_site_settings_model()
    }, {
        assert_debug_snapshot!(updated_settings);
    });
}

#[tokio::test]
#[serial]
async fn should_normalize_url_removing_trailing_slash() {
    configure_insta!();

    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    // 创建初始设置
    let _initial_settings =
        site_settings::Model::get_or_create(&boot.app_context.db, "http://localhost:5150")
            .await
            .unwrap();

    // 更新URL（带尾部斜杠）
    let url_with_slash = "https://example.com/";
    let settings =
        site_settings::Model::update_url(&boot.app_context.db, url_with_slash.to_string())
            .await
            .unwrap();

    // 验证尾部斜杠被去除
    assert_eq!(settings.site_url, "https://example.com");

    with_settings!({
        filters => cleanup_site_settings_model()
    }, {
        assert_debug_snapshot!(settings);
    });
}

#[tokio::test]
#[serial]
async fn should_reject_invalid_url_format() {
    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    // 创建初始设置
    let _initial_settings =
        site_settings::Model::get_or_create(&boot.app_context.db, "http://localhost:5150")
            .await
            .unwrap();

    // 尝试使用无效URL
    let invalid_url = "not-a-valid-url";
    let result =
        site_settings::Model::update_url(&boot.app_context.db, invalid_url.to_string()).await;

    assert!(result.is_err());
}

#[tokio::test]
#[serial]
async fn should_reject_url_without_http_scheme() {
    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    // 创建初始设置
    let _initial_settings =
        site_settings::Model::get_or_create(&boot.app_context.db, "http://localhost:5150")
            .await
            .unwrap();

    // 尝试使用没有 http/https 的URL
    let url_without_scheme = "example.com";
    let result =
        site_settings::Model::update_url(&boot.app_context.db, url_without_scheme.to_string())
            .await;

    assert!(result.is_err());
}

#[tokio::test]
#[serial]
async fn should_reject_empty_url() {
    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    // 创建初始设置
    let _initial_settings =
        site_settings::Model::get_or_create(&boot.app_context.db, "http://localhost:5150")
            .await
            .unwrap();

    // 尝试使用空URL
    let result = site_settings::Model::update_url(&boot.app_context.db, "".to_string()).await;

    assert!(result.is_err());
}

#[tokio::test]
#[serial]
async fn should_allow_both_http_and_https() {
    configure_insta!();

    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    // 创建初始设置
    let _initial_settings =
        site_settings::Model::get_or_create(&boot.app_context.db, "http://localhost:5150")
            .await
            .unwrap();

    // 测试 http
    let http_settings =
        site_settings::Model::update_url(&boot.app_context.db, "http://example.com".to_string())
            .await
            .unwrap();
    assert_eq!(http_settings.site_url, "http://example.com");

    // 测试 https
    let https_settings =
        site_settings::Model::update_url(&boot.app_context.db, "https://example.com".to_string())
            .await
            .unwrap();
    assert_eq!(https_settings.site_url, "https://example.com");

    with_settings!({
        filters => cleanup_site_settings_model()
    }, {
        assert_debug_snapshot!(https_settings);
    });
}

#[tokio::test]
#[serial]
async fn should_maintain_singleton_pattern() {
    configure_insta!();

    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    // 第一次获取或创建
    let settings1 =
        site_settings::Model::get_or_create(&boot.app_context.db, "http://localhost:5150")
            .await
            .unwrap();

    // 第二次获取或创建（应该返回已存在的记录）
    let settings2 =
        site_settings::Model::get_or_create(&boot.app_context.db, "http://different.com")
            .await
            .unwrap();

    // 应该是同一条记录
    assert_eq!(settings1.id, settings2.id);
    assert_eq!(settings1.site_url, settings2.site_url);

    with_settings!({
        filters => cleanup_site_settings_model()
    }, {
        assert_debug_snapshot!(settings2);
    });
}

fn cleanup_site_settings_model() -> Vec<(&'static str, &'static str)> {
    vec![
        (r"id: \d+,", "id: ID"),
        (
            r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?\+\d{2}:\d{2}",
            "DATE",
        ),
    ]
}
