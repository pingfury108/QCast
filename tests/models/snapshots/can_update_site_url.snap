---
source: tests/models/site_settings.rs
expression: updated_settings
---
Model {
    created_at: DATE,
    updated_at: DATE,
    id: ID
    site_url: "https://example.com",
}
